src/services/services.ts

```diff
@@ -4464,6 +4508,475 @@ namespace ts {
+                    // Also get all @types typings installed in visible node_modules directories
+                    for (const package of findPackageJsons(scriptPath)) {
+                        const typesDir = combinePaths(getDirectoryPath(package), "node_modules/@types");
+                        getCompletionEntriesFromDirectories(host, options, typesDir, result);
+                    }
+                }
+
+                return result;
+            }
+
+            function getCompletionEntriesFromDirectories(host: LanguageServiceHost, options: CompilerOptions, directory: string, result: ImportCompletionEntry[]) {
+                if (host.getDirectories && directoryProbablyExists(directory, host)) {
+                    for (let typeDirectory of host.getDirectories(directory)) {
+                        typeDirectory = normalizePath(typeDirectory);
+                        result.push(createCompletionEntryForModule(getBaseFileName(typeDirectory), ScriptElementKind.externalModuleName));
+                    }
+                }
+            }
+
+            function findPackageJsons(currentDir: string): string[] {
+                const paths: string[] = [];
+                let currentConfigPath: string;
+                while (true) {
+                    currentConfigPath = findConfigFile(currentDir, (f) => host.fileExists(f), "package.json");
+                    if (currentConfigPath) {
+                        paths.push(currentConfigPath);
+
+                        currentDir = getDirectoryPath(currentConfigPath);
+                        const parent = getDirectoryPath(currentDir);
+                        if (currentDir === parent) {
+                            break;
+                        }
+                        currentDir = parent;
+                    }
+                    else {
+                        break;
+                    }
+                }
+
+                return paths;
+            }
+
+
+            function enumerateNodeModulesVisibleToScript(host: LanguageServiceHost, scriptPath: string) {
+                const result: VisibleModuleInfo[] = [];
+
+                if (host.readFile && host.fileExists) {
+                    for (const packageJson of findPackageJsons(scriptPath)) {
+                        const package = tryReadingPackageJson(packageJson);
+                        if (!package) {
+                            return;
+                        }
+
+                        const nodeModulesDir = combinePaths(getDirectoryPath(packageJson), "node_modules");
+                        const foundModuleNames: string[] = [];
+
+                        if (package.dependencies) {
+                            addPotentialPackageNames(package.dependencies, foundModuleNames);
+                        }
+                        if (package.devDependencies) {
+                            addPotentialPackageNames(package.devDependencies, foundModuleNames);

```

Comment:
Yeah, I think both of those make sense. Not sure why I left them out originally 😆